### DISCUSSION TITLE
HeightMap from Texture

### DISCUSSION THREAD

**Discussion by Bodyclock-Games** [stride3d/stride#1899] [Q&A] [answered]
Hi All, making progress with porting my custom Unity terrain system to Stride. I have a custom vertex and I can create and draw the mesh. Now to add height. In the simplest Unity version I would read a value from a texture (texture.GetPixel(x,y)). This doesn't seem possible in Stride.
I see there is a HeightMap object (from which it appears I can create a collider). Is the workflow to create a HeightMap from a Texture and then use that to get heights and a collider (this would seem to help the slope and aspect processing I do).
If so, could someone point me to an example of getting a HeightMap from a Texture.

Edit: Or should I be using an Image (is Texture a GPU object?)

Edit: It's not going to win any awards but I'm getting there. Jus a float array from a texture. I'd still be curious about the use of Heightmaps and colliders though. Thanks.

![image](https://github.com/stride3d/stride/assets/49901560/43044a90-cde3-425d-b0bc-3a0afa2a4b35)

Cheers
Chris

**Comment by tebjan** [comment]
Hey, did you check here? https://github.com/Doprez/Awesome-Stride#terrain

**Reply by Bodyclock-Games** [reply]
I did thanks. I'll go back to it now I have a better understnding of how this all works in Stride.

**Comment by Idomeneas1970** [comment, marked as answer]
Hi Chris,
I tried a lot of different ways to read the pixel data from a texture as well, and then change them and add to a heightmap object; this worked for me the best, I hope it helps.
Another nice terrain sample is here: [https://github.com/SilentCLD/XenkoHMTerrain](https://github.com/SilentCLD/XenkoHMTerrain)
It is not using the terrain component setup, like some of the other tutorials, but they're all good to check out.
Cheers

```
            string assemblyPath = Path.GetDirectoryName(Assembly.GetExecutingAssembly().Location);
            string filename = AppContext.BaseDirectory + "Resources\\" + FileNameEditText.Text + ".bmp";
            Texture texture;
            using (var inStream = System.IO.File.OpenRead(filename))
                texture = Texture.Load(Game.GraphicsDevice, inStream, loadAsSRGB: true);

            Color[] heightValues = new Color[texture.Width * texture.Height];
            // Get the height information and put it in the array
            texture.GetData(Game.GraphicsContext.CommandList, heightValues);
            for (i = 0; i < TerrainWidth; i++)
            {
                for (j = 0; j < TerrainHeight; j++)
                {
                    index = (TerrainWidth * j) + i;
                    float height = Heightmap.HeightRange.X + heightValues[index].R *
                        (Heightmap.HeightRange.Y - Heightmap.HeightRange.X);
                    Heightmap.Shorts[index] = HeightmapUtils.ConvertToShortHeight(
                              short.MinValue, short.MaxValue, height);
                }
            }
```

**Reply by Bodyclock-Games** [reply]
Hi, thanks for that. That method will be useful. I'm making some progress. I found the best way to get height maps in (this is from Gaea) is 16bit raw and a Stride RawAsset. Then stream it into a byte array using Content.Stream and a method to go from bytes to floats.
The problem I have at the moment is adding vertex colours from a Texture. I have to go from Texture to Image to PixelBuffer and then use PixelBuffer.GetPixel(x,y) but i must have the color format messed up because it looks nothing like it should.
So tomorrow I'm going to try a variation of your method and use a PNG as a RawAsset and stream it into a System.Drawing.Image class and see where I go from there.
As I say, there has been some progress. This screenshot below is using a Texture over the whole mesh. My goal is to achieve the same with vertex colours.
![Progress](https://github.com/stride3d/stride/assets/49901560/8c3f342b-d0fe-464c-99b8-2b39c4b1f33c)

**Reply by Bodyclock-Games** [reply]
Just for your interest I got the vertex colours working quite simply in the end with a PNG added as a RawAsset and loaded to a System.Drawing.Bitmap from the Content.Stream method.

```
        private static Bitmap LoadColorMap(TerrainManager terrainManager, UrlReference urlReference)
        {
            return new(terrainManager.Content.OpenAsStream(urlReference));
        }
```

And then its a bit clunky because the GetPixel(x,y) method returns a System.Drawing.Color and that has to be translated to a Stride Color. But as shown below (renedered with vertex colors) it works. It could probably be streamlined but it's good enough for a proof of concept.

```
    System.Drawing.Color vColor = tileData.ColourMap.GetPixel(x, y);
    Color color = new(vColor.R, vColor.G, vColor.B, 1);
    vertices[vertexCount++] = new Vertex(position, normal, texCoord, color);
```

![image](https://github.com/stride3d/stride/assets/49901560/fbfa7a53-9019-4974-be02-ea14bfb0919c)
